lib/writer.js

```diff
@@ -206,9 +211,15 @@ Writer.prototype.connectTo = function(addr){
 Writer.prototype.close = function(fn){
   debug('start close');
 
+  this.pendingConns.each(function(conn){
+    conn.destroy();
+  });
+
   var n = this.conns.size();
-  this.conns.each(function(conn){
-    conn.end(function() {
+
+  if (fn && n === 0) fn();
+  else this.conns.each(function(conn){
+    conn.close(function() {
```

Comment:
As discussed offline, `close` is only for readers as it stops nsq from sending messeges.